### PR TITLE
Baseline work for launchpad / landingzone cmds

### PR DESCRIPTION
**Pull request by benc-uk**
This is a huge PR, but you could argue when we're still building the baseline code of the applications, PRs don't really work, anyhow...

This PR includes:

- Shared code for launchpad and landingzone commands in pkg/core
- The many MANY params and vars for those commands to run, grouped into an `Options` data structure
- Supporting changes in pkg/terraform to expand a directory to a set of --var-file options to pass to terraform
- Number supporting functions in pkg/azure for handling
  - Azure resource graph for finding resources
  - Azure CLI interface for accessing the current subscription and signed in user
  - Storage functions for finding storage accounts, getting access keys, and uploading blobs
- New console functions to support tfexec logging

The result is a:
- A functioning `rover launchpad run` command for the init, plan & deploy actions
- The detection of existing state and switching to enabling a backend if found. _Note. There is more to be done in this area_
- Uploading of local state when launchpad run is first run

### CODE REVIEW

**Comment by benc-uk** [comment]
image

**Review by sebastus — APPROVED** [review]
LGTM - nothing jumps out at me.

File: cmd/test.go [hunk 1]
```diff
@@ -26,8 +25,11 @@ Cobra is a CLI library for Go that empowers applications.
 This application is a tool to generate the needed files
 to quickly create a Cobra application.`,
 	Run: func(cmd *cobra.Command, args []string) {
-		auth, _ := terraform.Authenticate()
-		azure.FindStorageAccount("0", "demo", auth.SubscriptionID)
+		azure.UploadFileToBlob(
```
Comment (Nepomuceno) [inline cmd/test.go:28]: Hopefully this is not a valid subscription id 😉

**Review by Nepomuceno — APPROVED** [review]

**Review by Nepomuceno — APPROVED** [review]

File: cmd/test.go (hunk 1 above)
Comment (benc-uk) [reply cmd/test.go:28]: Hah, I think I'll remove this file altogether

File: pkg/azure/auth.go [hunk 2]
```diff
@@ -7,42 +7,17 @@
...
-		msiConfig := auth.NewMSIConfig()
-		azureAuthorizer, err := msiConfig.Authorizer()
-		cobra.CheckErr(err)
-
-		return azureAuthorizer
-	} else {
-		// This is fall through, and will fail if user is not logged in
-		azureAuthorizer, err := auth.NewAuthorizerFromCLI()
-		cobra.CheckErr(err)
+	// We defer everything to the Azure CLI
```
[the remark this answers is not in the document]
Comment (benc-uk) [reply pkg/azure/auth.go:18]: Yes, I confirmed this. It makes things a little simplier
There's a single source of the truth, for auth.

File: pkg/azure/storage.go [hunk 3]
```diff
@@ -1,35 +1,102 @@
...
 	client.Authorizer = GetAuthorizer()
 
-	accountPages, err := client.List(ctx)
+	keysRes, err := client.ListKeys(context.Background(), resGrp, accountName, storage.Kerb)
+	cobra.CheckErr(err)
+
+	return *(*keysRes.Keys)[0].Value
+}
+
+// ParseResourceID into subscription, resource group and name
```
[the remark this answers is not in the document]
Comment (benc-uk) [reply pkg/azure/storage.go:61]: Yeah good idea, but not utils, there should be a general azure.go in the azure package

File: pkg/core/action.go [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+
+const (
+	// ActionInit carries out a just init step and no real action
+	ActionInit Action = 1
+	// ActionPlan carries out a plan operation
+	ActionPlan Action = 2
+	// ActionDeploy carries out a plan AND apply operation
+	ActionDeploy Action = 3
+	// ActionDestroy carries out a destroy operation
+	ActionDestroy Action = 4
```
[the remark this answers is not in the document]
Comment (benc-uk) [reply pkg/core/action.go:24]: The current rover has init, plan, apply and destroy, what is very misleading to the user is "apply" actually runs terraform plan + terraform apply, so I'd rather use a different term like "deploy" to mean plan+apply

File: pkg/core/action.go [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+package core
+
+import (
+	"errors"
+	"strings"
+)
+
+type Action int
+
+const (
```
[the remark this answers is not in the document]
Comment (benc-uk) [reply pkg/core/action.go:16]: Good idea

File: pkg/azure/storage.go [hunk 6]
```diff
@@ -1,35 +1,102 @@
...
+	containerURL, _ := url.Parse(
+		fmt.Sprintf("https://%s.blob.core.windows.net/%s", accountName, blobContainer))
+
+	blobContainerURL := azblob.NewContainerURL(*containerURL, pipeline)
+	blobURL := blobContainerURL.NewBlockBlobURL(blobName)
+	file, err := os.Open(filePath)
+	cobra.CheckErr(err)
+
+	blobOptions := azblob.UploadToBlockBlobOptions{
+		BlockSize:   4 * 1024 * 1024,
```
[the remark this answers is not in the document]
Comment (benc-uk) [reply pkg/azure/storage.go:93]: I copied it from an example. I worked. I forgot about it and moved on!

File: pkg/azure/storage.go [hunk 7]
```diff
@@ -1,35 +1,102 @@
...
+		Parallelism: 16,
```
[the remark this answers is not in the document]
Comment (benc-uk) [reply pkg/azure/storage.go:94]: I think I'll remove these options and use whatever the internal defaults are, unless your state file is 100s of MB (which it should never be) I don't see these settings as important

File: pkg/core/options.go [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+}
+
+// NewOptionsFromCmd builds a Config from command flags
+func NewOptionsFromCmd(cmd *cobra.Command) Options {
+	launchPadMode := false
+	if cmd.Parent().Name() == "launchpad" {
+		launchPadMode = true
+	}
+
+	configPath, _ := cmd.Flags().GetString("config-path")
```
[the remark this answers is not in the document]
Comment (benc-uk) [reply pkg/core/options.go:42]: It's something I considered and we can discuss but the symphony.yaml file is effectively the config file that holds all these settings, see https://github.com/aztfmod/symphony/blob/master/symphony.yml and the `landingZonePath` `landingZoneConfigurationPath` `tfState` keys in there

**Pull request merged**